Sources/Public/BacktraceCrashLoopDetector.swift

```diff
@@ -0,0 +1,141 @@
+            a user’s defaults are stored locally on a single device,
+            and persisted for backup and restore.
+            To synchronize preferences and other data across a user’s connected devices,
+            use NSUbiquitousKeyValueStore instead.
+         */
+        guard let data = UserDefaults.standard.object(forKey: BacktraceCrashLoopDetector.plistKey) as? Data
+        else { return }
+
+        guard let array = try? PropertyListDecoder().decode([StartUpEvent].self, from: data)
+        else { return }
+                
+        startupEvents.append(contentsOf: array)
+        BacktraceLogger.debug("Events Loaded: \(startupEvents.count)")
+    }
+
+    @objc internal func saveEvents() {
+        let data = try? PropertyListEncoder().encode(startupEvents)
+        UserDefaults.standard.set(data, forKey: BacktraceCrashLoopDetector.plistKey)
+        BacktraceLogger.debug("Events Saved: \(startupEvents.count)")
+    }
+
+    @objc private func reportFilePath() -> String {
+        
+        // Crash Loop Detector considers all other Backtrace modules as potentially dangerous.
+        // Thats why it formats path to PLCrashReporter's report file itself,
+        // for not to use PLCrashReporter's APIs at all
+        
+        let bundleIDBT = Bundle.main.bundleIdentifier ?? ""
+        let appIDPath = bundleIDBT.replacingOccurrences(of: "/", with: "_")
+
+        let paths = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
+        let cacheDir = URL(fileURLWithPath: paths.isEmpty ? "" : paths[0])
+
+        let bundleIDPLCR = "com.plausiblelabs.crashreporter.data"
+        let crashReportDir = cacheDir.appendingPathComponent(bundleIDPLCR)
+                                     .appendingPathComponent(appIDPath)
+
+        let reportName = "live_report.plcrash"
+        let reportFullPath = crashReportDir.appendingPathComponent(reportName)
+                                           .absoluteString
+                                           .replacingOccurrences(of: "file://", with: "")
+        BacktraceLogger.debug("reportFullPath: \(reportFullPath)")
+        return reportFullPath
+    }
+
+    @objc private func hasCrashReport() -> Bool {
+        let exists = FileManager.default.fileExists(atPath: reportFilePath())
+        return exists
+    }
+
+    @objc private func addCurrentEvent() {
+        
+        var event = StartUpEvent(timestamp: Double(Date.timeIntervalSinceReferenceDate), isSuccessful: true)
+        event.isSuccessful = !hasCrashReport()
+        BacktraceLogger.debug("New Event: {timestamp:\(event.timestamp)--successful:\(event.isSuccessful)}")
+
+        startupEvents.append(event)
+        
+        if startupEvents.count > BacktraceCrashLoopDetector.consecutiveCrashesThreshold {
+            startupEvents.remove(at: 0)
+        }
+
+        BacktraceLogger.debug("Event Added: \(startupEvents.count)")
+    }
+
+    @objc private func badEventsCount() -> Int {
```

Comment:
badEventsCount doesn't describe also what it does - right now we're counting crashes. How about renaming it to startupCrashCounter() or something like that?

Reply:
we can also pass here as a parameter crashCounter and ask this function to check if we are in the crash loop. By doing that we do not need to iterate over the whole array (it's still small anyway and it doesn't cost as anyway, but the flow might be cleaner?)

Reply:
1. Agree, changed it to 'consecutiveEventsCount', as crashes should be consecutive to detect crash loop.
2. The function itself calculates only how many consecutive events happened during last N (threshold) launches. Detection of a crash loop is done inside detectCrashLoop() (it loads events, adds current event, saves events, calculates consecutive, compares them with threshold and returns true or false. 
3. I will elaborate documentation and comments a bit, so it will be less confusing :P